### ISSUE TITLE
[HOLD for payment 2023-12-28] [$500]  Mentioning an email with double @ sign results in an invalid email mention

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#28946]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open any chat
2. Type @ and select any email from suggestions
3. Move the cursor to the beginning
4. Add another @ sign
5. Hit the send button
6. Click on the mentioned email
## Expected Result: 
The valid email is mentioned
## Actual Result:
 Invalid email is mentioned
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.78-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/38435837/30ac83c8-54d6-4e7f-b85c-29c51703de6e

https://github.com/Expensify/App/assets/38435837/1a130972-b500-4fc1-96a7-dda7e465456c

https://github.com/Expensify/App/assets/38435837/6d52ffa5-8319-4dfc-8305-fabec4fed58d

https://github.com/Expensify/App/assets/38435837/8bdc424f-e2c4-4a13-b74f-372a7020ed7c

https://github.com/Expensify/App/assets/38435837/55a89672-3161-4291-93c3-c3675d3b7f76

https://github.com/Expensify/App/assets/38435837/d5cdfb6f-1ff6-460f-b607-3990adf6fb3f

https://github.com/Expensify/App/assets/38435837/4739b083-9660-4771-a34d-783e9d56e592

**Expensify/Expensify Issue URL:**
**Issue reported by:** @adeel0202 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1696445508377469

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016783de9809fcb5bd</li>
        <li>Upwork Job ID: 1710001713857413120</li>
        <li>Last Price Increase: 2023-12-07</li>
<li>Automatic offers: </li>
<ul>
<li>Victor-Nyagudi | Contributor | 28037832</li>
</ul></li>
    </ul>
</details>

**Comment by abdel-h66** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Mentions with Double leading @ redirect to the wrong route 

### What is the root cause of that problem?
To correctly calculate the display name or the login of the mentioned user, we only remove the leading @ 
using 

```js 
displayNameOrLogin = props.tnode.data.replace(CONST.UNICODE.LTR, '').slice(1);
```
Which causes the navigation link to contain the extra @

### What changes do you think we should make in order to solve the problem?
We can use the following regex to remove all the leading @ and correctly calculate the navigation link 

```js 
 displayNameOrLogin = props.tnode.data.replace(/^((CONST\.UNICODE\.LTR)*@)*/, '')
```

### What alternative solutions did you explore? (Optional)

N/A

**Comment by akamefi202** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Mentioning an email with double @ sign results in an invalid email mention.

### What is the root cause of that problem?
The root cause is replace regex of `userMentions` in `ExpensiMark` class of `expensify-common` repo.
https://github.com/Expensify/expensify-common/blob/bc275547d5e0e36cfd25169bdb6eb0ae3f505bf2/lib/ExpensiMark.js#L111-L120

Currently, it inputs the whole string inside `<mention-user>` tag if the string is matched with regex.
So if the string contains multiple `@` characters at the first, `<mention-user>` tag will include multiple `@` characters too.
```
@@test@test.com => <mention-user>@@test@test.com</mention-user>
```

That's why the app shows invalid email with `@` characters at first, on the mentioned email.

### What changes do you think we should make in order to solve the problem?
We need to update replace regex of `userMentions` to input only one `@` character inside `<mention-user>` tag.
Rest of `@` characters should stay outside `<mention-user>` tag.

```
name: 'userMentions',
regex: new RegExp(`[a-zA-Z0-9.!$%&+/=?^\`{|}-]?(@*?)(@${CONST.REG_EXP.EMAIL_PART})(?!((?:(?!<a).)+)?<\\/a>|[^<]*(<\\/pre>|<\\/code>))`, 'gm'),
replacement: (match, g1, g2) => {
    if (!Str.isValidMention(match)) {
        return match;
    }
    return `${g1}<mention-user>${g2}</mention-user>`;
},
```

As a result, the app will show valid email although the user inputs multiple `@` characters at first.
```
@@test@test.com => @<mention-user>@test@test.com</mention-user>
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by Victor-Nyagudi** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Mentioning an email with double @ sign results in an invalid email mention

### What is the root cause of that problem?

We're matching multiple `@` signs in the `regex` of user mentions rule in [`ExpensiMark`](https://github.com/Victor-Nyagudi/expensify-common/blob/bc275547d5e0e36cfd25169bdb6eb0ae3f505bf2/lib/ExpensiMark.js#L111-L120) i.e. the `@+` in the `@+${CONST.REG_EXP.EMAIL_PART}` section matches one or more `@` symbols.

```javascript
regex: new RegExp(`[a-zA-Z0-9.!$%&+/=?^\`{|}-]?@+${CONST.REG_EXP.EMAIL_PART}(?!((?:(?!<a).)+)?<\\/a>|[^<]*(<\\/pre>|<\\/code>))`, 'gm'),
``` 

### What changes do you think we should make in order to solve the problem?


I propose removing the `+` after the `@` symbol so only one `@` is matched. 

```javascript
regex: new RegExp(`[a-zA-Z0-9.!$%&+/=?^\`{|}-]?@${CONST.REG_EXP.EMAIL_PART}(?!((?:(?!<a).)+)?<\\/a>|[^<]*(<\\/pre>|<\\/code>))`, 'gm'),
```

This is what the regex will capture whenever there are multiple leading `@` symbols after this change. 

@@@`@johnsmith@gmail.com`

This behavior is also consistent with what we currently have when there are multiple leading `@` symbols for `@here` mentions. 

<details>

<summary>Demo after changes</summary>

https://github.com/Expensify/App/assets/79470910/16a2b217-d4ca-4a10-b258-43be95077d48

</details>

### What alternative solutions did you explore? (Optional)

None as of yet.

**Comment by Santhosh-Sellavel** [comment]
Thanks, @akamefi202 @Victor-Nyagudi Both of your proposals work.

I like your @Victor-Nyagudi as its cleaner and does the job. Wondering why + was added in the first place, are we missing something?

**Comment by Victor-Nyagudi** [comment]
I'm not sure either, @Santhosh-Sellavel. 

I looked through the merged PRs, and I didn't find any that added the `+`. It's possible it may have always been there from the beginning.

**Comment by abekkala** [comment]
@Santhosh-Sellavel do you have any further input for the proposals that were provided?

**Comment by Santhosh-Sellavel** [comment]
Asked [here](https://github.com/Expensify/expensify-common/pull/524/files#r1357339778) for clarifications regarding this [proposal](https://github.com/Expensify/App/issues/28946#issuecomment-1750185844).

**Comment by abekkala** [comment]
@getusha can you respond to @Santhosh-Sellavel question 

> Asked [here](https://github.com/Expensify/expensify-common/pull/524/files#r1357339778) for clarifications regarding this https://github.com/Expensify/App/issues/28946#issuecomment-1750185844.

**Comment by abekkala** [comment]
@Santhosh-Sellavel are we good to keep moving with that PR?

**Comment by abekkala** [comment]
@Santhosh-Sellavel is there an acceptable proposal here? Otherwise I can pull someone in to get a fix moving for this one

**Comment by Santhosh-Sellavel** [comment]
I guess we are good here as we have proposals already. I'll let post my comments soon.

**Comment by Santhosh-Sellavel** [comment]
@abekkala please add hold for the same, thanks!

**Comment by abekkala** [comment]
updated to HOLD

**Comment by abekkala** [comment]
@Santhosh-Sellavel Can you confirm we're still on HOLD for this one?

**Comment by abekkala** [comment]
still on hold

**Comment by abekkala** [comment]
still on hold

**Comment by abekkala** [comment]
On hold

**Comment by abekkala** [comment]
Still on hold

**Comment by Victor-Nyagudi** [comment]
@Santhosh-Sellavel The [PR](https://github.com/Expensify/expensify-common/pull/594) of the [issue we're holding for](https://github.com/Expensify/App/issues/21753) has been merged, and the version of `Expensify-Common` has been [bumped](https://github.com/Expensify/App/pull/31644) to the latest version which includes the changes. 

However, I can still reproduce the issue on latest `main` even after the PR was [deployed to staging](https://github.com/Expensify/App/pull/31644#issuecomment-1821929859).

<details>

<summary>Video of bug still reproducible after PR merge</summary>

https://github.com/Expensify/App/assets/79470910/b7c42a03-f685-42c6-8bd9-1da2e523b53a

</details>

My solution still holds even after the PR merge, and all tests are still passing. 

<details>

<summary>My changes after the PR merge</summary>

https://github.com/Expensify/App/assets/79470910/8797e5ac-46c9-46ac-82cf-77421a9813e1

</details>

Should we wait until the regression testing period is over to proceed? 

**Please Note:**

The fix from the merged PR wasn't showing in DEV, so I had to [change the commit hash to the one belonging to the PR](https://github.com/Expensify/App/pull/31541#pullrequestreview-1742332968) in `package.json` in order to see the changes.  

```diff
- "expensify-common": "git+ssh://git@github.com/Expensify/expensify-common.git#ee14b3255da33d2b6924c357f43393251b6dc6d2",
+ "expensify-common": "git+ssh://git@github.com/Expensify/expensify-common.git#d24d6fa86709a63b2ded8be0d3d76581d3aa08c5",
``` 

Everything works fine in staging. 

For reference, you could try pasting the following in the composer to see if it's highlighted correctly. 

```
@test.example+2@gmail.com@here
@here@test.example+2@gmail.com
@test.example+2@gmail.com@test.example+2@gmail.com
@here@here@gmail.com
@here@gmail.com@here
@here@here+1@gmail.com
@here+1@gmail.com@here
```

**Comment by abekkala** [comment]
I'm heading out for the Thanksgiving holiday and will be back bright and early Monday morning.
If something pops up that needs urgent attention please reapply label or ask in slack.

**Comment by abekkala** [comment]
@Santhosh-Sellavel can you address the comment above by @Victor-Nyagudi

**Comment by abekkala** [comment]
@Santhosh-Sellavel friendly bump here

**Comment by Santhosh-Sellavel** [comment]
@abekkala Sorry Hold label confuses. Can you take this off hold, please?

@Victor-Nyagudi Can you submit a formal proposal again, after the latest changes?

**Comment by Victor-Nyagudi** [comment]
@Santhosh-Sellavel I'll check on this tomorrow morning. 

If my proposal ends up being the same as the current one, can I just use that instead, leave a comment linking to my proposal titled "Updated", or just make a new identical one?

**Comment by Victor-Nyagudi** [comment]
Testing my solution.

**Comment by Victor-Nyagudi** [comment]
@Santhosh-Sellavel Tests well so far. 

I'm going to stick with my [original proposal](https://github.com/Expensify/App/issues/28946#issuecomment-1750185844).

**Comment by abekkala** [comment]
not overdue - @Santhosh-Sellavel can you review the proposal by Victor?

**Comment by Santhosh-Sellavel** [comment]
@Victor-Nyagudi [proposal](https://github.com/Expensify/App/issues/28946#issuecomment-1750185844) looks good to me!

**Comment by Santhosh-Sellavel** [comment]
@Victor-Nyagudi [proposal](https://github.com/Expensify/App/issues/28946#issuecomment-1750185844) looks good to me!

C+ Reviewed
🎀 👀 🎀

**Comment by abekkala** [comment]
@Victor-Nyagudi do you have a timeframe for the PR?

**Comment by amyevans** [comment]
Proposal looks good 👍

**Comment by Victor-Nyagudi** [comment]
Thank you for the offer.

I'll get to work on the PR. 

ETA the next 24-48 hours

**Comment by Victor-Nyagudi** [comment]
@Santhosh-Sellavel The PR is ready for review.

**Comment by amyevans** [comment]
We have the expensify-common PR merged, can we expect the App PR shortly @Victor-Nyagudi?

**Comment by Victor-Nyagudi** [comment]
Oh, sorry about that, @amyevans . I didn't know I should make an Expensify/App PR. 

Is it the one changing the expensify-common commit hash in Expensify/App to the latest commit in my PR?

**Comment by amyevans** [comment]
Exactly! 😄

**Comment by Victor-Nyagudi** [comment]
The docs put that under "Expensify only", so I thought only internal engineers should update expensify-common.

Nevertheless, it's late for me, so I'll make a PR first thing in the morning.

**Comment by Victor-Nyagudi** [comment]
@amyevans The [PR](https://github.com/Expensify/App/pull/33142) is ready for review. 

Commit hash in `package.json` has been updated. 

`license/snyk` tests are currently failing, but according to this [comment in a recently merged PR](https://github.com/Expensify/App/pull/32639#issuecomment-1852002668), it is a known issue that is currently being worked on. 

The PR was merged with those tests failing. 

cc: @Santhosh-Sellavel

**Comment by Victor-Nyagudi** [comment]
Friendly bump, @Santhosh-Sellavel, for a PR review. 

I just changed the commit hash, so no deep analysis of code required.

**Comment by abekkala** [comment]
@kadiealexander 
### STATUS
I'm ooo through Jan 03
This is at the finish line and last thing should be issuing payment once the regression period passes after PR is deployed to prod.

- @adeel0202 [$50] Issue Reporter
- @Victor-Nyagudi [$500 - if no regressions] Fix [Offer link](https://www.upwork.com/nx/wm/offer/28037832)
- @Santhosh-Sellavel [$500 - if no regressions] PR Review - **in NEWDOT, not Upwork**

**Comment by Santhosh-Sellavel** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [x]  [@Santhosh-Sellavel] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/expensify-common/pull/524
> * [x]  [@Santhosh-Sellavel] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/expensify-common/pull/524/files#r1357339778
> * [x]  [@Santhosh-Sellavel] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A 
> * [x]  [@Santhosh-Sellavel] Determine if we should create a regression test for this bug.  Yes
> * [ ]  [@Santhosh-Sellavel] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
> * [ ]  [@abekkala / @kadiealexander] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by Santhosh-Sellavel** [comment]
**Bug:**  Mentioning an email with double @ sign results in an invalid email mention
### Steps: 
1. Log into an account and open a report
2. Type '@' in the composer followed by the valid email of someone you've chatted with before. You can also optionally select one from the mentions' suggestions list.
3. Place the cursor at the beginning of the composer, type another '@' sign, and then send the message.
4. Verify that only the first '@' sign before the email and the email is highlighted e.g. @`@user@expensify.com`.
5. Repeat step 2 and then send the message.
6. Verify that a regular mention with one '@' sign works as usual.

**Comment by abekkala** [comment]
I'm back from ooo, unassigning @kadiealexander and creating regression test GH and payment today

**Comment by abekkala** [comment]
@Victor-Nyagudi  payment sent and contract ended - thank you! 🎉

**Comment by abekkala** [comment]
@adeel0202 payment offer sent in Upwork - once accepted, I can process payment

**Comment by abekkala** [comment]
@Santhosh-Sellavel  - [$500] to be made via New Dot

**Comment by abekkala** [comment]
@adeel0202 payment sent and contract ended - thank you! 🎉

**Comment by abekkala** [comment]
@Santhosh-Sellavel have you requested/received your payment in NewDot?

**Comment by Santhosh-Sellavel** [comment]
Requested on ND

**Comment by abekkala** [comment]
### PAYMENTS: 

Muhammad Adeel Nawaz paid ✅
Victor-Nyagudi paid ✅

Santosh - requested payment [$500] in ND

**Comment by JmillsExpensify** [comment]
$500 payment approved for @Santhosh-Sellavel based on [this comment](https://github.com/Expensify/App/issues/28946#issuecomment-1863414166).
